### PR TITLE
Added TP04 process

### PR DESCRIPTION
**Pull request by Cristiana959**
### Pull Request Overview

This pull request adds TP04(process).

### Testing Strategy

N/A

### TODO or Help Wanted

N/A

### Build

- [ ] Ran `npm build`.

### Author

Signed-off-by: Cristiana Andrei cristiana.andrei@wyliodrin.com

### CODE REVIEW

File: docs/tp/04.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+use nix::sys::wait::waitpid;
+use nix::sys::unistd::{fork, ForkResult, write};
+ 
+match unsafe{fork()} {
+   Ok(ForkResult::Parent { child, .. }) => {
+       println!("Continuing execution in parent process, new child has pid: {}", child);
+       waitpid(child, None).unwrap();
+   }
+   Ok(ForkResult::Child) => {
+       // Unsafe to use `println!` (or `unwrap`) here. See Safety.
```
Comment (alexandruradovici) [inline docs/tp/04.md:65]: Why?

Comment (alexandruradovici) [reply docs/tp/04.md:65]: Got it, this is valid only in multithreaded software, which is not the case here. You can use `println!`.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/04.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+```
+
+## Fonctions Unix
+### Les appels système dup et dup2
+#### dup
+L'appel système `dup()` crée une **copie** d'un descripteur de fichier.
+* Il utilise le descripteur inutilisé portant le numéro le plus bas pour le nouveau descripteur.
+* Si la copie est créée avec succès, les descripteurs de fichier d'origine et de copie peuvent être utilisés de manière interchangeable.
+* Ils se réfèrent tous les deux à la même description de fichier ouvert et partagent donc les indicateurs de décalage de fichier et d'état de fichier.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:79]:
```suggestion
* Ils se réfèrent tous les deux au même descripteur de fichier ouvert et partagent donc les indicateurs de décalage de fichier et d'état de fichier.
```

File: docs/tp/04.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+#### dup2
+L'appel système `dup2()` est similaire à `dup()` mais la différence fondamentale entre eux est qu'au lieu d'utiliser le descripteur de fichier inutilisé portant le numéro le plus bas, il utilise le numéro de descripteur **spécifié par l'utilisateur**.
+
+:::note
+Lors de l'appel de [dup2](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup2.html) dans Rust, les paramètres donnés doivent être l'ancien descripteur de fichier et le nouveau.
+:::
+
+### execvp()
+La fonction `execvp()` **remplace** l'image de processus actuelle par une nouvelle image de processus spécifiée par file. La nouvelle image est construite à partir d'un **fichier exécutable** standard appelé le nouveau fichier d'image de processus. Aucun retour n'est effectué car l'image de processus appelante est remplacée par la nouvelle image de processus.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:89]:
```suggestion
La fonction `execvp()` **remplace** l'image de processus actuelle par une nouvelle image de processus spécifiée par le fichier. La nouvelle image est construite à partir d'un **fichier exécutable** standard appelé le nouveau fichier d'image de processus. Aucun retour n'est effectué car l'image de processus appelante est remplacée par la nouvelle image de processus.
```

File: docs/tp/04.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+:::note
+```rust
+// utilisé pour trouver le descripteur de fichier 
+use std::os::unix::io::AsRawFd;
+```
+:::
+
+### Exercises
+1. Créer un programme de sorte que la contenu d'un fichier qui est recu comme premier argument soit **redirigée vers un fichier** reçu en tant que deuxième argument - une fonctionnalité similaire à l'utilitaire [cp](https://www.geeksforgeeks.org/cp-command-linux-examples/) de Linux.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:118]:
```suggestion
1. Créer un programme de sorte que le contenu d'un fichier qui est reçu comme premier argument soit **redirigée vers un fichier** reçu en tant que deuxième argument - une fonctionnalité similaire à l'utilitaire [cp](https://www.geeksforgeeks.org/cp-command-linux-examples/) de Linux.
```

File: docs/tp/04.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+2. Écrivez un programme qui alloue dynamiquement un String à l'aide de Box. Lisez un texte au clavier et ajoutez-le aux String allouée. Imprimez le String.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:119]:
```suggestion
2. Écrivez un programme qui alloue dynamiquement un String à l'aide de Box. Lisez un texte depuis le clavier et ajoutez-le à la String allouée. Imprimez le String.
```

**Review by alexandruradovici — APPROVED** [review]

**Pull request merged**